### PR TITLE
fix: transactional isolation for syncAccount (#31)

### LINKED ISSUE #31: fix: transactional isolation for syncAccount to prevent partial data loss
## Summary

`syncAccount(_:)` deletes all old positions before rebuilding them from fresh API data. If the rebuild fails mid-way (e.g., `upsertAsset` throws due to a SwiftData fetch error), the `save()` call on the all-accounts-failed path persists the deletions and any partial inserts — turning a transient error into permanent data loss.

## Origin

CodeRabbit review on PR #29 (critical finding). The delete-before-rebuild pattern pre-dates PR #29, but the fetch helper error propagation introduced in that PR creates a new throwing path inside the rebuild loop.

## Acceptance Criteria

- `VERIFY:` A failed `syncAccount` rebuild does not persist deleted positions or partial inserts
- `VERIFY:` Successful rebuilds still persist correctly
- `VERIFY:` `lastSyncError` is still set and persisted on failure (regression from #6 must not return)

## Approach Options

1. **Scratch ModelContext** — perform per-account mutations on a child/scratch `ModelContext`, merge to the main context only on success
2. **Deferred delete** — collect positions to delete, but only call `modelContext.delete()` after the rebuild succeeds
3. **Rollback on failure** — wrap mutations in a transaction and rollback on error (if SwiftData supports it)

## Files

- `Sources/Portu/Sync/SyncEngine.swift` — `syncAccount(_:)` (lines ~51-107)

## Complexity Hint

- **Subsystems:** SyncEngine (single module)
- **Structural changes:** Yes — requires transactional isolation or deferred mutation pattern

### PR DESCRIPTION
**Pull request by rube-de**
## Summary

- Fix data loss when `syncAccount` rebuild fails mid-loop. Previously, positions were deleted before rebuilding from API data; if `upsertAsset` threw, the `save()` on the all-accounts-failed path persisted the deletions and partial inserts.
- Restructured into a two-phase build/commit pattern: positions and tokens are staged in memory first, and old positions are only deleted (and new ones inserted) after the build phase succeeds.
- Added regression tests via a `StubProvider` and an `upsertAssetOverride` closure injected into `SyncEngine` to force a mid-rebuild upsert failure.

Closes #31

## Test plan

- [x] New test `failed rebuild preserves existing positions` — fails on the old code, passes on the new
- [x] New test `successful rebuild replaces positions` — regression guard for the success path
- [x] Existing test `lastSyncError persisted when all syncable accounts fail` still passes
- [x] Full Xcode test suite passes (261 tests, 42 suites)
- [x] SPM package tests pass (PortuCore, PortuNetwork, PortuUI)
- [x] SwiftLint clean (no new warnings)

### CODE REVIEW

File: Sources/Portu/Sync/SyncEngine.swift [hunk 1]
```diff
@@ -100,7 +101,21 @@ final class SyncEngine: @unchecked Sendable {
...
+        }
+
+        for (position, tokens) in staged {
+            position.account = account
             modelContext.insert(position)
+            for token in tokens {
+                token.position = position
+                modelContext.insert(token)
+            }
         }
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Sync/SyncEngine.swift]: Answered: Kept explicit `modelContext.insert` for each token (rather than relying on `position.tokens = tokens`) for symmetry with the snapshot-and-defensive-ordering pattern other reviewers asked for in nearby threads — assigning relationships before insertion can rely on inverse hooks firing for unmanaged objects, which is the SwiftData footgun the new code now avoids by inserting first and then assigning the relationship.

File: Sources/Portu/Sync/SyncEngine.swift [hunk 2]
```diff
@@ -4,8 +4,9 @@ import PortuCore
 import PortuNetwork
 import SwiftData
 
+/// Non-final to allow test subclass override of upsertAsset (ThrowingSyncEngine).
 @MainActor
-final class SyncEngine: @unchecked Sendable {
+class SyncEngine: @unchecked Sendable {
     private let modelContext: ModelContext
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Sync/SyncEngine.swift]: Answered: Deliberate testability choice. The class doc-comment now documents the subclass contract (subclasses must be `@MainActor`-isolated to preserve `@unchecked Sendable`), and `upsertAsset` carries an inline note explaining why it is not `private`. Closure injection is a viable future refactor but would touch the public init signature; current approach keeps the existing surface stable while enabling the test seam.

File: Sources/Portu/Sync/SyncEngine.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Sync/SyncEngine.swift:148]: Fixed: Applied both defensive hardenings — `Array(account.positions)` snapshot before the delete loop, and insert-before-relationship-assignment in the staged loop. Scratch `ModelContext` is intentionally out of scope for #31 (which targets build-phase isolation only); commit-phase resilience would require a child context and is now noted as a tracked-separately concern in the commit-phase comment.

File: Sources/Portu/Sync/SyncEngine.swift [hunk 3]
```diff
@@ -4,8 +4,9 @@ import PortuCore
 import PortuNetwork
 import SwiftData
 
+/// Non-final to allow test subclass override of upsertAsset (ThrowingSyncEngine).
 @MainActor
-final class SyncEngine: @unchecked Sendable {
+class SyncEngine: @unchecked Sendable {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Sync/SyncEngine.swift]: Answered: Same as the copilot thread on line 10 — deliberate testability choice. The class doc-comment now documents the `@MainActor` subclass contract, and `upsertAsset` carries an inline note explaining why it is not `private`. Closure injection is a viable future refactor; current approach keeps the existing public init stable.

File: Sources/Portu/Sync/SyncEngine.swift [hunk 4]
```diff
@@ -63,12 +64,13 @@ final class SyncEngine: @unchecked Sendable {
...
 
-        // Delete stale positions from previous sync
-        for position in account.positions {
-            modelContext.delete(position)
-        }
+        // ── Build phase: stage positions and tokens without inserting them ──
+        // If upsertAsset throws mid-loop, no positions have been deleted or
+        // inserted, so the account's data survives the failed attempt (#31).
+        // (Asset upserts may insert new Assets as a side effect; these are
+        // harmless shared reference data reused by future syncs.)
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Sync/SyncEngine.swift]: Fixed: Expanded the build-phase comment to spell out the asset side effects (`upsertAsset` may insert new Assets, `updateAssetMetadata` may mutate existing ones, both flushed by the subsequent save) and named the self-healing invariant — the 3-tier dedup hierarchy plus append-only key backfill in `upsertAsset` prevents Asset accumulation across repeated partial failures.

File: Sources/Portu/Sync/SyncEngine.swift [hunk 5]
```diff
@@ -63,12 +64,13 @@ final class SyncEngine: @unchecked Sendable {
...
+        var staged: [(Position, [PositionToken])] = []
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Sync/SyncEngine.swift]: Fixed: Replaced the anonymous tuple with named `StagedPosition` / `StagedToken` structs at file scope. They store plain value types only (no @Model instances) so the build phase no longer auto-registers PositionToken rows via SwiftData's relationship tracking on the resolved Asset — a leak the CodeRabbit-suggested orphan-token assertion actually surfaced.

File: Sources/Portu/Sync/SyncEngine.swift [hunk 6]
```diff
@@ -100,7 +101,21 @@ final class SyncEngine: @unchecked Sendable {
             }
 
             position.netUSDValue = net
+            staged.append((position, tokens))
+        }
+
+        // ── Commit phase: all DTOs succeeded — replace positions ──
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Sync/SyncEngine.swift:113]: Fixed: Wrapped the iteration in `Array(account.positions)` to snapshot before the cascade-delete mutation. Also flagged independently by CodeRabbit on the same lines — same fix covers both threads.

File: Sources/Portu/Sync/SyncEngine.swift [hunk 7]
```diff
@@ -100,7 +101,21 @@ final class SyncEngine: @unchecked Sendable {
             }
 
             position.netUSDValue = net
+            staged.append((position, tokens))
+        }
+
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Sync/SyncEngine.swift:112]: Fixed: Added a comment at the commit-phase boundary documenting the limitation — the build-phase isolation in #31 covers the upsertAsset-throws path; commit-phase resilience would require a child ModelContext and is tracked separately.

File: Tests/PortuTests/SyncEngineTests.swift [hunk 8]
```diff
@@ -1,22 +1,27 @@
...
 import PortuCore
+import PortuNetwork
 import SwiftData
 import Testing
 
 @MainActor
 struct SyncEngineTests {
-    private func makeTestContext() throws -> (ModelContext, SyncEngine) {
+    /// Use a fresh ModelContext per test — container.mainContext shares thread-local
+    /// state across tests which causes SIGTRAP when multiple containers exist.
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/SyncEngineTests.swift]: Fixed: Rewrote to identify the actual hazard — reusing `container.mainContext` across tests, not the existence of multiple containers.

File: Tests/PortuTests/SyncEngineTests.swift [hunk 9]
```diff
@@ -169,6 +174,92 @@ struct SyncEngineTests {
         #expect(allAssets.count == 1)
     }
 
+    // MARK: - Transactional Isolation (#31)
+
+    /// Issue #31: If upsertAsset throws mid-rebuild, existing positions
+    /// must survive — the deferred delete must not have executed.
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/SyncEngineTests.swift]: Fixed: Replaced "deferred delete" with "commit-phase delete" to match the production code's block comment terminology.

File: Tests/PortuTests/SyncEngineTests.swift [hunk 10]
```diff
@@ -169,6 +174,92 @@ struct SyncEngineTests {
...
+    @Test func `failed rebuild preserves existing positions`() async throws {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/SyncEngineTests.swift:181]: Fixed: Added `failed rebuild on first upsert preserves existing positions` covering the `throwAfter: 0` boundary case where the build phase stages nothing and the commit phase is a no-op.

File: Tests/PortuTests/SyncEngineTests.swift [hunk 11]
```diff
@@ -169,6 +174,92 @@ struct SyncEngineTests {
...
+            #expect(error == .allAccountsFailed)
+        }
+
+        // Verify via fresh context: original position must survive
+        let freshContext = ModelContext(context.container)
+        let accounts = try freshContext.fetch(FetchDescriptor<Account>())
+        let fetched = try #require(accounts.first)
+
+        #expect(fetched.positions.count == 1, "Original position must survive failed rebuild")
+        #expect(fetched.positions[0].tokens.first?.asset?.symbol == "OLD")
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/SyncEngineTests.swift]: Fixed: Replaced `fetched.positions[0]` with `.first?.tokens.first?.asset?.symbol` to avoid relying on positional ordering. Also added `allTokens.count == 1` to assert no orphan tokens (CodeRabbit suggested the same in the review body — and it surfaced a real leak that prompted the build-phase restructure to plain-value staging).

File: Tests/PortuTests/SyncEngineTests.swift [hunk 12]
```diff
@@ -194,6 +285,27 @@ struct SyncEngineTests {
             category: category, isVerified: isVerified)
     }
 
+    private func makeMockContext(balances: [PositionDTO]) throws -> (ModelContext, SyncEngine) {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/SyncEngineTests.swift:332]: Answered: Both helpers now share `makeModelContext()` for the schema/container/context boilerplate. The remaining 3-line difference is just the engine type they construct (SyncEngine vs ThrowingSyncEngine) — collapsing further would either require a generic helper or inlining setup into each test, which trades duplication for readability friction. Left as-is.

File: Tests/PortuTests/SyncEngineTests.swift [hunk 13]
```diff
@@ -242,3 +354,44 @@ private final class MockSecretStore: SecretStore, @unchecked Sendable {
         store.removeValue(forKey: key.rawKey)
     }
 }
+
+// MARK: - StubProvider
+
+private struct StubProvider: PortfolioDataProvider {
+    var capabilities: ProviderCapabilities {
+        ProviderCapabilities()
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/SyncEngineTests.swift:415]: Fixed: Converted `StubProvider` from `struct` to `actor` with `nonisolated var capabilities` to match the `PortfolioDataProvider` actor pattern used by `ZapperProvider`, `ExchangeProvider`, and the existing `MockProvider` in `PortuNetworkTests`.

File: Tests/PortuTests/SyncEngineTests.swift [hunk 14]
```diff
@@ -242,3 +354,44 @@ private final class MockSecretStore: SecretStore, @unchecked Sendable {
...
+
+/// Test subclass that forces upsertAsset to throw after N successful calls,
+/// simulating a mid-rebuild failure for transactional isolation testing.
+@MainActor
+private class ThrowingSyncEngine: SyncEngine {
+    let throwAfter: Int
+    private var upsertCount = 0
+
+    init(modelContext: ModelContext, providerFactory: ProviderFactory, throwAfter: Int) {
+        self.throwAfter = throwAfter
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/SyncEngineTests.swift]: Fixed: Added a doc comment to `throwAfter` clarifying the semantics — `throwAfter: 0` throws on the first call, `throwAfter: 1` lets the first succeed and throws on the second, etc. Kept the name to avoid touching helper signatures across the test setup.

File: Tests/PortuTests/SyncEngineTests.swift [hunk 15]
```diff
@@ -242,3 +354,44 @@ private final class MockSecretStore: SecretStore, @unchecked Sendable {
...
+        super.init(modelContext: modelContext, providerFactory: providerFactory)
+    }
+
+    override func upsertAsset(from dto: TokenDTO) throws -> Asset {
+        upsertCount += 1
+        if upsertCount > throwAfter {
+            throw SyncTestError.forcedUpsertFailure
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Tests/PortuTests/SyncEngineTests.swift]: Fixed: Added a "single-use" note to the `ThrowingSyncEngine` doc comment explicitly stating that `upsertCount` does not reset between `sync()` calls, so each test must create a fresh instance.

**Comment by rube-de** [comment]
> **Actionable comments posted: 1**

Fixed: Applied all three nitpicks.
- **Orphan PositionToken assertion** (Tests/PortuTests/SyncEngineTests.swift:177-221) — added `allTokens.count == 1`. This actually surfaced a real leak: setting `asset` on `PositionToken` during init auto-registers the token via SwiftData's relationship tracking, defeating the build-phase isolation. Restructured staging to use plain-value structs (`StagedPosition` / `StagedToken`) so no @Model objects are constructed before the commit phase.
- **Build-phase Asset-mutation comment** (Sources/Portu/Sync/SyncEngine.swift:67-105) — expanded the comment to spell out that `upsertAsset` may also run `updateAssetMetadata` on existing Assets, and named the self-healing dedup invariant.
- **Subclass MainActor contract doc** (Sources/Portu/Sync/SyncEngine.swift:7-9) — added a class-level note that subclasses must remain `@MainActor`-isolated to preserve `@unchecked Sendable`.

**Comment by rube-de** [comment]
> ## CI Review

Partially addressed:
- **Fixed (#3, upsertAsset doc):** Added inline comment at the declaration explaining why it is not `private` (overridden by ThrowingSyncEngine in tests).
- **Answered (#1, no logging in catch):** The `catch` block in `sync()` (line 34) is pre-existing and not modified by this PR. Adding structured logging there is a valid follow-up but out of scope for #31 (build-phase isolation in `syncAccount`).
- **Answered (#2, `error.localizedDescription` leak):** Same — line 35 is pre-existing. Sanitizing error storage is a valid follow-up but out of scope for #31.

**Comment by rube-de** [comment]
## Review compliance check (cycle 1)

Addressed all 22 review items (16 inline threads + 4 review bodies + 2 issue comments) across 5 reviewers. Code changes pushed in [`39bc01a`](https://github.com/Chalet-Labs/portu/commit/39bc01a276047848c4ffe3a664528005256f6ea1).

### Status by category

| Category | Count |
|---|---|
| Fixed | 13 (12 inline + 1 review body) |
| Discussion-Answered | 5 (4 inline + 1 review body) |
| Resolved (silent — non-actionable) | 4 (2 review bodies + 2 issue comments) |

### Per-reviewer breakdown

| Reviewer | Items | Outcome |
|---|---|---|
| @claude | 14 | 11 Fixed, 2 Discussion-Answered (`final` removal), 1 Discussion-Answered partial (CI Review body — pre-existing items deferred) |
| @coderabbitai | 3 | 2 Fixed (inline + review-body nitpicks), 1 Resolved (walkthrough issue comment) |
| @copilot-pull-request-reviewer | 2 | 1 Discussion-Answered (`final` removal), 1 Resolved (PR-overview body) |
| @gemini-code-assist | 2 | 1 Discussion-Answered (token-relationship style), 1 Resolved (summary body) |
| @codacy-production | 1 | 1 Resolved ("0 issues" success body) |

### Highlights

The CodeRabbit-suggested orphan PositionToken assertion actually surfaced a real leak — assigning a managed Asset on `PositionToken` during init auto-registers the token via SwiftData's relationship tracking. Restructured the build phase to stage with plain-value structs (`StagedPosition` / `StagedToken`) so no @Model objects are constructed before the commit phase. The new boundary test (`failed rebuild on first upsert preserves existing positions`, `throwAfter: 0`) covers the empty-staged case.

### Out-of-scope deferrals (tracked separately)

The following pre-existing items in `sync()` (not modified by this PR) remain valid concerns but are out of #31 scope:
- Per-account sync failures lack structured logging (claude CI Review #1)
- `error.localizedDescription` may include sensitive URL data (claude CI Review #2)
- Commit-phase resilience would require a child `ModelContext` (CodeRabbit + claude on `SyncEngine.swift:106-119`)

The two architectural calls — restoring `final` with closure injection (3 reviewers) and using a scratch `ModelContext` (CodeRabbit) — were declined as conflicting with the PR's deliberate testability and scope choices, with class-level documentation added to communicate the contract.

**Comment by rube-de** [comment]
> **Actionable comments posted: 1**
> 
> **Nitpick comments (3)** — `Tests/PortuTests/SyncEngineTests.swift:177-221`, `Sources/Portu/Sync/SyncEngine.swift:67-105`, `Sources/Portu/Sync/SyncEngine.swift:7-9`...

Fixed: All three CodeRabbit nitpicks applied and verified against `39bc01a`:
- **Orphan `PositionToken` assertion** (`Tests/PortuTests/SyncEngineTests.swift:223-224`): `allTokens.count == 1` after failed rebuild — catches regressions where staged tokens auto-register through managed relationships
- **Build-phase comment** (`Sources/Portu/Sync/SyncEngine.swift:68-79`): updated to clarify that `upsertAsset` may insert new Asset rows or run `updateAssetMetadata` on existing ones, and explains why those side-effects are self-healing via the 3-tier dedup hierarchy
- **`@unchecked Sendable` subclass contract** (`Sources/Portu/Sync/SyncEngine.swift:7-8`): doc-comment now states all subclasses must remain `@MainActor`-isolated to preserve the unchecked claim

**Comment by rube-de** [comment]
> ## CI Review
> 
> **Profile**: agent | **Findings**: 16 (0 critical, 3 high, 8 medium, 5 low)
> ...
> ### Findings Not in Diff
> - **[medium] error-handling** `Sources/Portu/Sync/SyncEngine.swift:34`
> - **[medium] security** `Sources/Portu/Sync/SyncEngine.swift:35`
> - **[low] quality** `Sources/Portu/Sync/SyncEngine.swift:128`

Addressed:
- **Fixed (#3, `upsertAsset` accessibility):** `/// Internal (not private) so ThrowingSyncEngine can override it in tests.` doc-comment added at `Sources/Portu/Sync/SyncEngine.swift:156` — verified at `39bc01a`.
- **Answered (#1, error logging):** Pre-existing `catch` block at line 35 has no structured logging. Acknowledged as out-of-scope for this PR — this PR addresses the data-loss bug in #31. Deferred to a follow-up.
- **Answered (#2, `localizedDescription` security):** Same catch block — `error.localizedDescription` may expose URL/host details for network errors. Acknowledged as out-of-scope for this PR; deferred to a follow-up focused on error sanitization.

File: Sources/Portu/Sync/SyncEngine.swift [hunk 16]
```diff
@@ -4,8 +4,10 @@ import PortuCore
 import PortuNetwork
 import SwiftData
 
+/// Non-final to allow test subclass override of upsertAsset (ThrowingSyncEngine).
+/// Any subclass must remain @MainActor-isolated to preserve the @unchecked Sendable contract.
 @MainActor
-final class SyncEngine: @unchecked Sendable {
+class SyncEngine: @unchecked Sendable {
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Sync/SyncEngine.swift]: Fixed: restored `SyncEngine` to `final class`; replaced `ThrowingSyncEngine` subclass with `upsertAssetOverride` closure seam (commit 5611bef). `@unchecked Sendable` is now scoped to a sealed final class — no subclass can violate the isolation invariant.

File: Sources/Portu/Sync/SyncEngine.swift [hunk 17]
```diff
@@ -8,6 +8,8 @@ import SwiftData
 final class SyncEngine: @unchecked Sendable {
     private let modelContext: ModelContext
     private let providerFactory: ProviderFactory
+    /// Test seam: overrides upsertAsset when set; nil in production.
+    var upsertAssetOverride: ((TokenDTO) throws -> Asset)?
```
[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Sync/SyncEngine.swift]: Fixed: updated PR description to replace all `ThrowingSyncEngine` subclass references with the `upsertAssetOverride` closure injection approach.

[the remark this answers is not in the document]
Comment (rube-de) [reply Sources/Portu/Sync/SyncEngine.swift]: Fixed: wrapped `upsertAssetOverride` declaration and its `if let override` dispatch in `#if DEBUG` / `#endif` (commit 7205afe). The property is absent from Release builds entirely.

**Comment by rube-de** [comment]
> ## CI Review
> 
> **Profile**: single | **Findings**: 1 (0 critical, 0 high, 1 medium, 0 low)...

Fixed: restored `SyncEngine` to `final class` and replaced the subclassing approach with an `upsertAssetOverride` closure seam (commits 5611bef, 7205afe). The `@unchecked Sendable` contract is now scoped to a sealed class and the seam is `#if DEBUG`-guarded.

**Comment by rube-de** [comment]
> ## CI Review
> 
> **Profile**: single | **Findings**: 1 — `upsertAssetOverride` not guarded by `#if DEBUG`...

Fixed: wrapped `upsertAssetOverride` declaration and dispatch in `#if DEBUG` / `#endif` (commit 7205afe). Absent from Release builds.

**Comment by rube-de** [comment]
> 🧹 Nitpick comments (1) — `Tests/PortuTests/SyncEngineTests.swift` (lines 227–262): add orphan-token assertion...

Fixed: added `allTokens.count == 1` assertion to the `throwAfter:0` isolation test (commit 7205afe), matching the invariant already present in the `throwAfter:1` variant.
